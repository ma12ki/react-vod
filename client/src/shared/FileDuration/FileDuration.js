import React from 'react';
import PropTypes from 'prop-types';

export const FileDuration = ({ duration }) => {
    const { hours, minutes, seconds } = splitDuration(duration);
    const secondsText = zeroFill(seconds);
    const minutesText = zeroFill(minutes);
    const hoursText = hours;

    const finalText = [hoursText, minutesText, secondsText]
        .join(':');

    return (
        <span>{finalText}</span>
    );
};

FileDuration.propTypes = {
    duration: PropTypes.number.isRequired,
};

const splitDuration = (seconds) => {
    const secondRemainder = seconds % 60;
    const minutes = (seconds - secondRemainder) / 60;
    const minuteRemainder = minutes % 60;
    const hours = (minutes - minuteRemainder) / 60;

    return {
        hours,
        minutes: minuteRemainder,
        seconds: Math.round(secondRemainder),
    };
};

const zeroFill = (num) => `00${num}`.substr(-2, 2);

export default FileDuration;
