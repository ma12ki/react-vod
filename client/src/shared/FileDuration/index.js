export { default as FileDuration } from './FileDuration';
