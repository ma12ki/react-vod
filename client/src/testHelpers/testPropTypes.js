import React from 'react';

// this has some limitations:
// https://stackoverflow.com/questions/41916992/why-does-my-jest-test-of-react-proptypes-break-when-using-multiple-unacceptable/41916993#41916993
export const testPropTypes = (component, propName, acceptableValues, unacceptableValues, otherProps) => {
    console.error = jest.fn();
    const _test = (testValues, expectError) => {
        for (let propValue of testValues) {
            console.error.mockClear();
            React.createElement(component, {...otherProps, [propName]: propValue});
            try {
                expect(console.error).toHaveBeenCalledTimes(expectError ? 1 : 0);
            } catch (err) {
                console.log(`${propName} expected to ${expectError ? '' : 'NOT'} fail for value ${propValue}`);
                throw err;
            }
        }
    };
    _test(acceptableValues || [], false);
    _test(unacceptableValues || [], true);
};
