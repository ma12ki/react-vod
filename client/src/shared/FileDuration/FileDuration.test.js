import React from 'react';
import { shallow } from 'enzyme';

import { testPropTypes } from '../../testHelpers';
import { FileDuration } from './FileDuration';

const makeWrapper = (duration = 12345) => shallow(<FileDuration duration={duration} />);

describe('FileDuration', () => {

    it('matches the snapshot', () => {
        const wrapper = makeWrapper();

        expect(wrapper).toMatchSnapshot();
    });

    it('requires "duration" number prop', () => {
        testPropTypes(
            FileDuration,
            'duration',
            [0, 1, 20, 300, 4000, 50000],
            ['123', null, undefined]
        );
    });

    it('properly renders time', () => {
        [
            { input: 0, output: '0:00:00'},
            { input: 28, output: '0:00:28'},
            { input: 180, output: '0:03:00'},
            { input: 3599, output: '0:59:59'},
            { input: 3600, output: '1:00:00'},
            { input: 3721, output: '1:02:01'},
        ].forEach(({ input, output }) => {
            const wrapper = makeWrapper(input);

            expect(wrapper.text()).toEqual(output);
        });
    });

});