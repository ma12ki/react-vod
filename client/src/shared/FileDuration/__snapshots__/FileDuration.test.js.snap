// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileDuration matches the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <span>
    3:25:45
</span>,
  "nodes": Array [
    <span>
      3:25:45
</span>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Unknown
        duration={12345}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "duration": 12345,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <span>
          3:25:45
</span>,
        "_debugID": 2,
        "_renderedOutput": <span>
          3:25:45
</span>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Unknown
    duration={12345}
/>,
}
`;
